### PR TITLE
feat: implement source refs for helm set-file (Beta) (#13220)

### LINKED ISSUE #13220: Cannot use file from other source as Helm --set-file param
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

In Application using multiple sources reference of one source is not substituted in helm.fileParameters entries. It works (at least it seems) for the helm.valueFiles property

**To Reproduce**

Using this Application definition:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: kratos
  namespace: argocd
spec:
  project: ory
  destination:
    name: ''
    namespace: auth
    server: 'https://kubernetes.default.svc'
  sources:
    - repoURL: 'https://k8s.ory.sh/helm/charts'
      targetRevision: 0.30.0
      chart: kratos
      helm:
        valueFiles:
          - $myRepo/kratos/k8s/overlay/development/kratos-values.yaml

        fileParameters:
          - name: identity
            path: $myRepo/kratos/k8s/overlay/development/identity_schemas/identity.schema.json
    - repoURL: git@my-private:repo.git
      targetRevision: development
      ref: myRepo
```
gives this error in the UI:

```
rpc error: code = Unknown desc = Manifest generation error (cached): `helm template . --name-template kratos --namespace auth --kube-version 1.25 --set-file identity=/tmp/47de275d-f0c3-419a-8382-9628f1674103/kratos/kratos/k8s/overlay/development/identity_schemas/identity.schema.json --values .kratos/k8s/overlay/development/kratos-values.yaml --api-versions acid.zalan.do/v1 --api-versions acid.zalan.do/v1/OperatorConfiguration --api-versions acid.zalan.do/v1/PostgresTeam --api-versions acid.zalan.do/v1/postgresql --api-versions admissionregistration.k8s.io/v1 --api-versions admissionregistration.k8s.io/v1/MutatingWebhookConfiguration --api-versions admissionregistration.k8s.io/v1/ValidatingWebhookConfiguration --api-versions apiextensions.k8s.io/v1 --api-versions apiextensions.k8s.io/v1/CustomResourceDefinition --api-versions apiregistration.k8s.io/v1 --api-versions apiregistration.k8s.io/v1/APIService --api-versions apps/v1 --api-versions apps/v1/ControllerRevision --api-versions apps/v1/DaemonSet --api-versions apps/v1/Deployment --api-versions apps/v1/ReplicaSet --api-versions apps/v1/StatefulSet --api-versions argoproj.io/v1alpha1 --api-versions argoproj.io/v1alpha1/AppProject --api-versions argoproj.io/v1alpha1/Application --api-versions argoproj.io/v1alpha1/ApplicationSet --api-versions authzed.com/v1alpha1 --api-versions authzed.com/v1alpha1/SpiceDBCluster --api-versions auto.gke.io/v1 --api-versions auto.gke.io/v1/AllowlistedV2Workload --api-versions auto.gke.io/v1/AllowlistedWorkload --api-versions auto.gke.io/v1alpha1 --api-versions auto.gke.io/v1alpha1/AllowlistedWorkload --api-versions autoscaling/v1 --api-versions autoscaling/v1/HorizontalPodAutoscaler --api-versions autoscaling/v2 --api-versions autoscaling/v2/HorizontalPodAutoscaler --api-versions autoscaling/v2beta2 --api-versions autoscaling/v2beta2/HorizontalPodAutoscaler --api-versions batch/v1 --api-versions batch/v1/CronJob --api-versions batch/v1/Job --api-versions certificates.k8s.io/v1 --api-versions certificates.k8s.io/v1/CertificateSigningRequest --api-versions cloud.google.com/v1 --api-versions cloud.google.com/v1/BackendConfig --api-versions cloud.google.com/v1beta1 --api-versions cloud.google.com/v1beta1/BackendConfig --api-versions coordination.k8s.io/v1 --api-versions coordination.k8s.io/v1/Lease --api-versions core.strimzi.io/v1beta2 --api-versions core.strimzi.io/v1beta2/StrimziPodSet --api-versions discovery.k8s.io/v1 --api-versions discovery.k8s.io/v1/EndpointSlice --api-versions events.k8s.io/v1 --api-versions events.k8s.io/v1/Event --api-versions flowcontrol.apiserver.k8s.io/v1beta1 --api-versions flowcontrol.apiserver.k8s.io/v1beta1/FlowSchema --api-versions flowcontrol.apiserver.k8s.io/v1beta1/PriorityLevelConfiguration --api-versions flowcontrol.apiserver.k8s.io/v1beta2 --api-versions flowcontrol.apiserver.k8s.io/v1beta2/FlowSchema --api-versions flowcontrol.apiserver.k8s.io/v1beta2/PriorityLevelConfiguration --api-versions getambassador.io/v1beta1 --api-versions getambassador.io/v1beta1/RateLimit --api-versions getambassador.io/v1beta2 --api-versions getambassador.io/v1beta2/Filter --api-versions getambassador.io/v1beta2/FilterPolicy --api-versions getambassador.io/v1beta2/RateLimit --api-versions getambassador.io/v2 --api-versions getambassador.io/v2/AuthService --api-versions getambassador.io/v2/ConsulResolver --api-versions getambassador.io/v2/DevPortal --api-versions getambassador.io/v2/Filter --api-versions getambassador.io/v2/FilterPolicy --api-versions getambassador.io/v2/Host --api-versions getambassador.io/v2/KubernetesEndpointResolver --api-versions getambassador.io/v2/KubernetesServiceResolver --api-versions getambassador.io/v2/LogService --api-versions getambassador.io/v2/Mapping --api-versions getambassador.io/v2/Module --api-versions getambassador.io/v2/RateLimit --api-versions getambassador.io/v2/RateLimitService --api-versions getambassador.io/v2/TCPMapping --api-versions getambassador.io/v2/TLSContext --api-versions getambassador.io/v2/TracingService --api-versions getambassador.io/v3alpha1 --api-versions getambassador.io/v3alpha1/AuthService --api-versions getambassador.io/v3alpha1/ConsulResolver --api-versions getambassador.io/v3alpha1/DevPortal --api-versions getambassador.io/v3alpha1/Filter --api-versions getambassador.io/v3alpha1/FilterPolicy --api-versions getambassador.io/v3alpha1/Host --api-versions getambassador.io/v3alpha1/KubernetesEndpointResolver --api-versions getambassador.io/v3alpha1/KubernetesServiceResolver --api-versions getambassador.io/v3alpha1/Listener --api-versions getambassador.io/v3alpha1/LogService --api-versions getambassador.io/v3alpha1/Mapping --api-versions getambassador.io/v3alpha1/Module --api-versions getambassador.io/v3alpha1/RateLimit --api-versions getambassador.io/v3alpha1/RateLimitService --api-versions getambassador.io/v3alpha1/TCPMapping --api-versions getambassador.io/v3alpha1/TLSContext --api-versions getambassador.io/v3alpha1/TracingService --api-versions hub.gke.io/v1 --api-versions hub.gke.io/v1/Membership --api-versions internal.autoscaling.gke.io/v1alpha1 --api-versions internal.autoscaling.gke.io/v1alpha1/CapacityRequest --api-versions kafka.strimzi.io/v1alpha1 --api-versions kafka.strimzi.io/v1alpha1/KafkaTopic --api-versions kafka.strimzi.io/v1alpha1/KafkaUser --api-versions kafka.strimzi.io/v1beta1 --api-versions kafka.strimzi.io/v1beta1/KafkaTopic --api-versions kafka.strimzi.io/v1beta1/KafkaUser --api-versions kafka.strimzi.io/v1beta2 --api-versions kafka.strimzi.io/v1beta2/Kafka --api-versions kafka.strimzi.io/v1beta2/KafkaBridge --api-versions kafka.strimzi.io/v1beta2/KafkaConnect --api-versions kafka.strimzi.io/v1beta2/KafkaConnector --api-versions kafka.strimzi.io/v1beta2/KafkaMirrorMaker --api-versions kafka.strimzi.io/v1beta2/KafkaMirrorMaker2 --api-versions kafka.strimzi.io/v1beta2/KafkaRebalance --api-versions kafka.strimzi.io/v1beta2/KafkaTopic --api-versions kafka.strimzi.io/v1beta2/KafkaUser --api-versions migration.k8s.io/v1alpha1 --api-versions migration.k8s.io/v1alpha1/StorageState --api-versions migration.k8s.io/v1alpha1/StorageVersionMigration --api-versions monitoring.googleapis.com/v1 --api-versions monitoring.googleapis.com/v1/ClusterPodMonitoring --api-versions monitoring.googleapis.com/v1/ClusterRules --api-versions monitoring.googleapis.com/v1/GlobalRules --api-versions monitoring.googleapis.com/v1/OperatorConfig --api-versions monitoring.googleapis.com/v1/PodMonitoring --api-versions monitoring.googleapis.com/v1/Rules --api-versions monitoring.googleapis.com/v1alpha1 --api-versions monitoring.googleapis.com/v1alpha1/ClusterPodMonitoring --api-versions monitoring.googleapis.com/v1alpha1/ClusterRules --api-versions monitoring.googleapis.com/v1alpha1/GlobalRules --api-versions monitoring.googleapis.com/v1alpha1/OperatorConfig --api-versions monitoring.googleapis.com/v1alpha1/PodMonitoring --api-versions monitoring.googleapis.com/v1alpha1/Rules --api-versions networking.gke.io/v1 --api-versions networking.gke.io/v1/ManagedCertificate --api-versions networking.gke.io/v1/ServiceAttachment --api-versions networking.gke.io/v1beta1 --api-versions networking.gke.io/v1beta1/FrontendConfig --api-versions networking.gke.io/v1beta1/ManagedCertificate --api-versions networking.gke.io/v1beta1/ServiceAttachment --api-versions networking.gke.io/v1beta1/ServiceNetworkEndpointGroup --api-versions networking.gke.io/v1beta2 --api-versions networking.gke.io/v1beta2/ManagedCertificate --api-versions networking.k8s.io/v1 --api-versions networking.k8s.io/v1/Ingress --api-versions networking.k8s.io/v1/IngressClass --api-versions networking.k8s.io/v1/NetworkPolicy --api-versions node.k8s.io/v1 --api-versions node.k8s.io/v1/RuntimeClass --api-versions nodemanagement.gke.io/v1alpha1 --api-versions nodemanagement.gke.io/v1alpha1/UpdateInfo --api-versions oathkeeper.ory.sh/v1alpha1 --api-versions oathkeeper.ory.sh/v1alpha1/Rule --api-versions policy/v1 --api-versions policy/v1/PodDisruptionBudget --api-versions rbac.authorization.k8s.io/v1 --api-versions rbac.authorization.k8s.io/v1/ClusterRole --api-versions rbac.authorization.k8s.io/v1/ClusterRoleBinding --api-versions rbac.authorization.k8s.io/v1/Role --api-versions rbac.authorization.k8s.io/v1/RoleBinding --api-versions redis.redis.opstreelabs.in/v1beta1 --api-versions redis.redis.opstreelabs.in/v1beta1/Redis --api-versions redis.redis.opstreelabs.in/v1beta1/RedisCluster --api-versions scheduling.k8s.io/v1 --api-versions scheduling.k8s.io/v1/PriorityClass --api-versions snapshot.storage.k8s.io/v1 --api-versions snapshot.storage.k8s.io/v1/VolumeSnapshot --api-versions snapshot.storage.k8s.io/v1/VolumeSnapshotClass --api-versions snapshot.storage.k8s.io/v1/VolumeSnapshotContent --api-versions snapshot.storage.k8s.io/v1beta1 --api-versions snapshot.storage.k8s.io/v1beta1/VolumeSnapshot --api-versions snapshot.storage.k8s.io/v1beta1/VolumeSnapshotClass --api-versions snapshot.storage.k8s.io/v1beta1/VolumeSnapshotContent --api-versions storage.k8s.io/v1 --api-versions storage.k8s.io/v1/CSIDriver --api-versions storage.k8s.io/v1/CSINode --api-versions storage.k8s.io/v1/CSIStorageCapacity --api-versions storage.k8s.io/v1/StorageClass --api-versions storage.k8s.io/v1/VolumeAttachment --api-versions storage.k8s.io/v1beta1 --api-versions storage.k8s.io/v1beta1/CSIStorageCapacity --api-versions v1 --api-versions v1/ConfigMap --api-versions v1/Endpoints --api-versions v1/Event --api-versions v1/LimitRange --api-versions v1/Namespace --api-versions v1/Node --api-versions v1/PersistentVolume --api-versions v1/PersistentVolumeClaim --api-versions v1/Pod --api-versions v1/PodTemplate --api-versions v1/ReplicationController --api-versions v1/ResourceQuota --api-versions v1/Secret --api-versions v1/Service --api-versions v1/ServiceAccount --api-versions warden.gke.io/v1 --api-versions warden.gke.io/v1/Audit --include-crds` failed exit status 1: Error: failed parsing --set-file data: open /tmp/47de275d-f0c3-419a-8382-9628f1674103/kratos/kratos/k8s/overlay/development/identity_schemas/identity.schema.json: no such file or directory
14 minutes ago (Thu Apr 13 2023 17:01:24 GMT+0200)
```

Notice this:

```
--set-file identity=/tmp/47de275d-f0c3-419a-8382-9628f1674103/kratos/kratos/k8s/overlay/development/identity_schemas/identity.schema.json 
--values .kratos/k8s/overlay/development/kratos-values.yaml
```

in `--values` param the `$myRepo` is substituted with a `.` and in `--set-file` with `/tmp/47de275d-f0c3-419a-8382-9628f1674103/kratos`.

**Expected behavior**

The files in the gite repo should be used to set the variables in Helm

**Screenshots**

image

**Version**

```shell
argocd: v2.6.7+5bcd846
  BuildDate: 2023-03-23T14:57:27Z
  GitCommit: 5bcd846fa16e4b19d8f477de7da50ec0aef320e5
  GitTreeState: clean
  GoVersion: go1.18.10
  Compiler: gc
  Platform: linux/amd64
```

### PR DESCRIPTION
**Pull request by cardoe**
In multi-source applications helm values files can come from other repos if you use the ref syntax but it is not possible for the --set-file param to take files from other repos via the ref syntax. This extends the behavior to cover set-file parameters as well as the values files. fixes #13220, refs #17822

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by cardoe** [comment]
This still needs testing added. The documentation pieces are in #17822, which I will rebase this PR on top once that lands. I'll readily admit I'm not a Go programmer so probably did some things wrong from an idiomatic standpoint. Hoping another contributor is willing to help me get this over the finish line.

**Comment by cardoe** [comment]
I've labeled this as beta since it's for multi-source applications which themselves are beta.

File: reposerver/repository/repository_test.go
Comment (morremeyer) [inline reposerver/repository/repository_test.go:2246]: @cardoe Do you think it makes sense to add another assertion here to verify that the `password` parameter is set?

Comment (cardoe) [reply reposerver/repository/repository_test.go:2246]: Yeah. Struggling a little to know which array position in the Manifests I need to look at.

Comment (morremeyer) [reply reposerver/repository/repository_test.go:2246]: You might want to check out [slices.IndexFunc](https://pkg.go.dev/golang.org/x/exp/slices), if there’s specific values you can match on that could work.

If you want, I can take a look next Monday.

Comment (cardoe) [reply reposerver/repository/repository_test.go]: Well it turns out I was setting the wrong file parameter. I thought I had set the one that didn't base64 the input. Correcting that helped. I think this is a better test now. But again I'll readily admit I'm not a Go programmer so your input is definitely appreciated.

I have been able to take this branch and build the container and use it for a project where I had a reference to a helm file parameter and it went from erroring to setting the value correctly.

**Comment by cardoe** [comment]
I've set the documentation update to say the feature is available from ArgoCD 2.12+ because I'm assuming this won't make it into 2.11 since that's already in RC status.

**Comment by cardoe** [comment]
So I pushed once to confirm the test failed. And the most recent test confirms the test passes.

File: reposerver/repository/repository_test.go
Comment (morremeyer) [reply reposerver/repository/repository_test.go]: I can't open a PR against your branch, not sure why. 

I added one commit at https://github.com/morremeyer/argo-cd/commit/aa090e2112b3e4cf88bffa6c07e5abab8cdc9e74 that uses `slices.IndexFunc` to check against all manifests, not just one. 

Did not run the full test suite against it yet, feel free to add it to this PR.

**Comment by morremeyer** [comment]
@cardoe I can't open a PR against your branch, not sure why.

I added one commit at https://github.com/morremeyer/argo-cd/commit/aa090e2112b3e4cf88bffa6c07e5abab8cdc9e74 that uses slices.IndexFunc to check against all manifests, not just one.

Did not run the full test suite against it yet, feel free to add it to this PR.

**Comment by cardoe** [comment]
Thanks @morremeyer. I've squashed your change in since you prefixed it with `fixup!` and rebased the branch. Thanks for doing that!

**Review by morremeyer — APPROVED** [review]

**Comment by cardoe** [comment]
I've got this patch built against 2.12.x and have been using it myself.

**Comment by iulianivan** [comment]
Hello, 
Any chance of moving forward with this PR?
It's a very useful feature in my opinion.
Thanks

File: docs/user-guide/helm.md
Comment (morremeyer) [inline docs/user-guide/helm.md:240]: Looking at this again, I think the whole warning can be removed - the documentation is versioned.

Comment (cardoe) [reply docs/user-guide/helm.md]: Removed.

**Comment by cardoe** [comment]
I've rebased the branch and removed the documentation chunk that @morremeyer asked.

**Comment by cardoe** [comment]
I've rebased once again and attempted to add a test that checks for failure not just the happy path.

**Comment by morremeyer** [comment]
We have now built ArgoCD with this PR and are running it on our infrastructure for this week to validate the feature and observe performance.

I'll report back about that on Friday.

**Comment by morremeyer** [comment]
This has worked very well in our setup, from my point of view this PR is ready to be merged.

Thanks @cardoe for contributing this, we'll be very happy to see it released!

@crenshaw-dev Thanks for your feedback in the Contributor sync, as discussed then, we tested it in our infrastructure with success.

**Review by morremeyer — APPROVED** [review]

**Comment by phyllipecesar** [comment]
Could we merge and release it?

**Comment by cardoe** [comment]
Like @morremeyer I'm successfully running this patch in my infra. Overall this increases test coverage. I know the codecov says there's some partials and missed tests but that's cause the calls to `pathutil.ResolveValueFilePathOrUrl()` aren't tested for failure. But that function is tested elsewhere so I didn't think it'd be worth the increased cycles.

**Comment by fllaca** [comment]
hi @morremeyer, the PR looks in a good shape 🤩, do you know if there is any news on this feature being accepted and released?

**Comment by cardoe** [comment]
At this time we're waiting on @crenshaw-dev to review it. I've pinged him about it on Slack.

**Comment by gopaljayanthi** [comment]
this is a useful feature for configmap files in non-helm repos, please review it and merge so it is available to users at large. if user acceptance test is required, then i am willing to offer my services  to test the feature.

**Review by agaudreault — APPROVED** [review]
LGTM

**Review by crenshaw-dev — APPROVED** [review]
Thanks @cardoe for the PR, @agaudreault for the review, and others for the testing!

**Comment by spy16** [comment]
Will this be part of v2.12.0 ??

**Comment by morremeyer** [comment]
@spy16 No, it was merged after the first RC was made, so it will not appear in v2.12.0.

We can expect it with v2.13.0 if I remember the flow correctly.

**Pull request merged**
